// datdot-ui-timeline-days/index.js
const debug = require('debug')
const bel = require('bel')
const csjs = require('csjs-inject')
const { format, getDate, getMonth, getYear, getDaysInMonth, isToday } = require('date-fns')

module.exports = datdot_ui_timeline_days

function datdot_ui_timeline_days({data = null, style}, protocol) {
    const name = 'ui-timeline-days'
    const log = debug(name)
    const date = new Date()
    const today = getDate(date)
    const sendToParent = protocol(function receive (message) { log(message) })

    if ( data === null ) {
        // if no data
        var count = getMonth(date) // initial month
        var month = format(date, 'MMMM') // get capitalization month
        var year = getYear(date) // initial year
        var days = getDaysInMonth(date) // initial days
    } else {
        // if data is loaded
        var { from, count, month, year, days } = data
    }

    const is_today = (d) => isToday(new Date(year, count, d)) 

    
    const el = bel`<section role="timeline-days" class=${style}></section>`

    for (let d = 1; d <= days; d++ ) {
        let date = format(new Date(year, count, d), 'd MMMM  yyyy, EEEE')
        let setDate = format(new Date(year, count, d), 'yyyy-M-d')
        
        if (d < today && count <= new Date().getMonth() && year <= new Date().getFullYear() ) {
            var day = bel`<div role="button" class="${css['timeline-day']} ${css['disabled-date']}" tabindex="-1" aria-today=${is_today(d)} aria-label="${date}" data-date="${setDate}">${d}</div>`
        } else if ( count < new Date().getMonth() && year <= new Date().getFullYear() ) {
            var day = bel`<div role="button" class="${css['timeline-day']} ${css['disabled-date']}" tabindex="-1" aria-today=${is_today(d)} aria-label="${date}" data-date="${setDate}">${d}</div>`
        } else if ( year < new Date().getFullYear()) {
            var day = bel`<div role="button" class="${css['timeline-day']} ${css['disabled-date']}" tabindex="-1" aria-today=${is_today(d)} aria-label="${date}" data-date="${setDate}">${d}</div>`
        }
         else {
            var day = bel`<div role="button" class="${css['timeline-day']} ${is_today(d) ? `${css.today} ${css['date-selected']}` : ''}" tabindex="-1" aria-today=${is_today(d)} aria-label="${date}" aria-selected="${is_today(d)}" data-date="${setDate}" onclick=${(el) => onclick( el.target, setDate )}>${d}</div>`
        }


        el.append(day)
    }
    
    return el

    function onclick(target, date) {
        const message = { from: from ? from : ui, flow: ui, type: 'click', body: date, count, month, year, days }
        const logger = log.extend(`day> ${message.body}`)

        const children = [...el.children]
        children.forEach( btn => {
            btn.classList.remove(css['date-selected'])
            btn.removeAttribute('aira-selected')
        })

        target.classList.add(css['date-selected'])
        target.setAttribute('aria-selected', true)

        logger('date selected', date);
        logger('send', message)

        return sendToParent(message)
    }
}

const css = csjs`
.timeline-day {
    text-align: center;
    padding: 8px;
    cursor: pointer;
    transition: color 0.25s, background-color 0.25s ease-in-out;
}
.timeline-day:hover {
    color: #fff;
    background-color: #212121;
}
.timeline-day:focus, .timeline-day:active {
    outline: dotted 1px #c9c9c9;
}
.today {
    background-color: #f2f2f2;
}
.date-selected {
    color: #fff;
    background-color: #000;
}
.disabled-date {
    color: #BBBBBB;
    cursor: default;
}
.disabled-date:hover {
    color: #BBBBBB;
    background: none; 
}
`
