// datdot-ui-datepicker/index.js
const bel = require('bel')
const csjs = require('csjs-inject')
const debug = require('debug')
const monthResult = require('month-result')
const { isPast, isFuture } = require('date-fns')
// widgets
const calendarDays = require('../datdot-ui-calendar-days')
const calendarMonth = require('../datdot-ui-calendar-month')
const svg = require('svg')

module.exports = datepicker

function datepicker({name = 'datepicker', month1, month2, status = 'cleared'}, protocol) {
    const widget = 'ui-datepicker'
    const log = debug(`${widget}/${name}`)
    log('ready')
    const sendToParent = protocol( receive )
    let name1 = 'calendar1'
    let name2 = 'calendar2'
    let count = month1[1]
    const recipients = {}
    let value = {}
    let first, second
    // elements
    let cal1 = calendarDays({name: name1, month: month1[1], days: month1[2], year: month1[0], status }, calendarDaysProtocol(name1))
    let cal2 = calendarDays({name: name2, month: month2[1], days: month2[2], year: month2[0], status }, calendarDaysProtocol(name2))
    const weekList= ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']
    const title1 = calendarMonth({from: name, getDate: new Date(month1[0], month1[1]), view: 'datepicker-range-days'}, calendarMonthProtocol)
    const title2 = calendarMonth({from: name, getDate: new Date(month2[0], month2[1]), view: 'datepicker-range-days'}, calendarMonthProtocol)
    const iconPrev = svg( { css: `${css.icon} ${css['icon-prev']}`, path: './src/assets/arrow-left.svg' } )
    const iconNext = svg( { css: `${css.icon} ${css['icon-next']}`, path: './src/assets/arrow-right.svg' } )
    const prevMonth  = bel`<button role="button" aria-label="Previous month" class="${css.btn} ${css.prev}" onclick=${triggerPreviousMonth}>${iconPrev}</button>`
    const nextMonth = bel`<button role="button" aria-label="Next month" class="${css.btn} ${css.next}" onclick=${triggerNextMonth}>${iconNext}</button>`
    const container = bel`<div class=${css['calendar-container']}></div>`

    container.append( calendarView(title1, cal1), calendarView(title2, cal2) )

    const element = bel`
    <div class=${css.datepicker}>
        <div class=${css["calendar-header"]}>${prevMonth}${nextMonth}</div>
        ${container}
    </div>`

    return element

    function calendarView (title, calendar) {
        const cal = bel`<div class=${css.calendar}>${title}${makeWeekDays()}${calendar}</div>`
        return cal
    }

    function calendarMonthProtocol (send) {
        return function receiveMonth( message ) {
            log(message)
        }
    }

    function triggerPreviousMonth () {
        count -= 2
        const prevCal1 = monthResult(count)
        const prevCal2 = monthResult(count + 1)
        const message1 = { from: name, type: 'change', body: prevCal1}
        const message2 = { from: name, type: 'change', body: prevCal2}
        const ca1 = recipients[name1](message1)
        const ca2 = recipients[name2](message2)
        const month1Title = calendarMonth({from: name, getDate: new Date(prevCal1.year, prevCal1.count), view: 'datepicker-range-days'}, calendarMonthProtocol)
        const month2Title = calendarMonth({from: name, getDate: new Date(prevCal2.year, prevCal2.count), view: 'datepicker-range-days'}, calendarMonthProtocol)
        container.innerHTML = ''
        container.append( calendarView(month1Title, ca1), calendarView(month2Title, ca2) )

        const pastMonth = value.first ? isPast(new Date(prevCal1.year, prevCal1.count, prevCal1.days)) : undefined
        if (pastMonth) return forwardMessage(name, {from: name, type: 'color-range-from-start'})
    }
    function triggerNextMonth () {
        count += 2
        const nextCal1 = monthResult(count)
        const nextCal2 = monthResult(count + 1)
        const message1 = { from: name, type: 'change', body: nextCal1}
        const message2 = { from: name, type: 'change', body: nextCal2}
        const ca1 = recipients[name1](message1)
        const ca2 = recipients[name2](message2)
        const month1Title = calendarMonth({from: name, getDate: new Date(nextCal1.year, nextCal1.count), view: 'datepicker-range-days'}, calendarMonthProtocol)
        const month2Title = calendarMonth({from: name, getDate: new Date(nextCal2.year, nextCal2.count), view: 'datepicker-range-days'}, calendarMonthProtocol)
        container.innerHTML = ''
        container.append( calendarView(month1Title, ca1), calendarView(month2Title, ca2) )
        
        const nextMonth = value.first ? isFuture(new Date(nextCal1.year, nextCal1.count, nextCal1.days)) : undefined
        if (nextMonth) return forwardMessage(name, {from: name, type: 'color-range-to-end'})
    }

    function makeWeekDays () {
        const el = bel`<section class=${css['calendar-weekday']} role="weekday"></section>`
        weekList.map( w => {
            let div = bel`<div class=${css['calendar-week']} role="week">${w.slice(0 ,1)}</div>`
            el.append(div)
        })
        return el
    }

    function calendarDaysProtocol (name) {
        return send => {
            recipients[name] = send
            return receive
        }
    }

    function receive (message) {
        const {from, type, body} = message
        log(`<= ${type} <= ${from}`) 
        if (type === 'value/first') return notifyAndStoreFirst(from, body)
        if (type === 'value/second') return notifyParent(from, body)
        if (type === 'selecting-second') return notifyOtherCalenderSelectingLast(from)
        if (type === 'cleared') return clearOther( from === name1 ? name2 : name1)
        else return forwardMessage(from, message)
    }

    function filterOutRecipients (list) {
        const receivers = Object.keys(recipients)
                               .filter( name => !list.includes(name))
                               .map(name => recipients[name])
        return receivers
    }

    function forwardMessage (from, message) {
        const list = [from]
        const receivers = filterOutRecipients(list)
        broadcast(receivers, message)
    }

    function clearOther (from) {
        const send = recipients[from]
        return send({ from: name, type: 'clear'})
    }

    function notifyParent (from, body) {
        value.second = body
        const message = {from: widget, type: 'value', body: value} 
        log(`notify parent: ${widget} ${body}`)
        sendToParent({from, flow: widget, type:'value/second', body})
        return sendToParent(message)
    }

    function notifyOtherCalenderSelectingLast (from) {
        if (from === name1) {
            const send = recipients[name2]
            const message  = {from: name, type: 'color-from-start'}
            log(message)
            return send( message )
        }

        if (from === name2) {
            const send = recipients[name1]
            const message = {from: name, type: 'color-to-end'}
            log(message)
            return send(message)
        }
    }
    
    function notifyAndStoreFirst (from, body) {
        value.first = body
        const list = [from]
        const receivers = filterOutRcipients(list)
        const type = from === name1 ? 'first-selected-by-startcal' : 'first-selected-by-endcal'
        sendToParent({from, flow: widget, type: 'value/first', body})
        return broadcast( receivers, {from: name, type}) 
    }

    function filterOutRcipients (list) {
        const receivers = Object.keys(recipients)
                                .filter( name => !list.includes(name))
                                .map( name => recipients[name])
        return receivers
    }

    function broadcast (receivers, message) {
        const {from, type, body} = message
        for ( let i = 0, len = receivers.length; i < len; i++) {
            const send = receivers[i]
            log('boradcast =>', `${body} => ${from}`)
            send(message)
        }
    }

}


const css = csjs`
.datepicker {
    position: relative;
    max-width: 510px;
}
.datepicker-body {
    display: grid;
    grid-template-rows: auto;
    grid-template-columns: repeat(2, 210px);
    grid-gap: 35px;
}
.btn {
    background: none;
    border: none;
    border-radius: 50px;
    width: 30px;
    height: 30px;
    padding: 0;
    transition: background-color 0.3s ease-in-out;
    cursor: pointer;
}
.btn:active, .btn:hover {
    background-color: #C9C9C9;
}
.prev {}
.next {}
.icon svg path {
    transition: stroke 0.25s ease-in-out;
}
.icon-prev {}
.icon-next {}
.calendar-header {
    position: absolute;
    z-index: 9;
    display: flex;
    justify-content: space-between;
    width: 100%;
}
.calendar-container {
    display: flex;
}
.calendar-weekday {
    display: grid;
    grid-template-rows: 30px;
    grid-template-columns: repeat(7, minmax(30px, auto));
    justify-items: center;
    font-size: 12px;
}
.calendar-week {
    
}
.calendar {
    margin-left: 30px;
}
.title {
    font-size: 18px;
    text-align: center;
}
`
